config/confighttp/confighttp.go

```diff
@@ -131,7 +138,44 @@ func (hss *HTTPServerSettings) ToServer(handler http.Handler) *http.Server {
 		co := cors.Options{AllowedOrigins: hss.CorsOrigins}
 		handler = cors.New(co).Handler(handler)
 	}
+	if hss.EnableDecompression {
+		handler = httpDecompressionHandler(handler)
+	}
 	return &http.Server{
 		Handler: handler,
 	}
 }
+
+// httpDecompressionHandler is a middleware that helps offload the task of handling compressed
+// HTTP requests by identifying the compression format in the "Content-Encoding" header and re-writing
+// request body so that the handlers further in the chain can work on decompressed data.
+// It supports gzip and deflate/zlib compression.
+func httpDecompressionHandler(handler http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		switch r.Header.Get("Content-Encoding") {
+		case "gzip":
+			gr, err := gzip.NewReader(r.Body)
+			if err != nil {
+				// TODO: OTLP expects error of type google.rpc.Status
```

Comment:
In the [specs](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/protocol/otlp.md#bad-data), it's noted that the error should be proto encoded, do we expect this to be common across all usage of this middleware(via `HTTPServerSettings`)? If it's not, this can become a problem.

Reply:
So I've refactored the code a bit to accommodate a custom error handler function. Hope I'm going in the right direction. 
https://github.com/open-telemetry/opentelemetry-collector/pull/1581/files#diff-cdf3ee5da427d239900ca510e7bc43dcR32

However, I'm not sure how we can make this configurable via the settings. Would something like this work?

```
http:
   decompression:
      enabled: true
      error_format: otlp
```

Reply:
Also, I just noticed that the contrib library is using `HTTPServerSettings.ToServer`. I suppose it will be difficult to have functions that work for all use-cases. Eg: [signalfxreceiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/master/receiver/signalfxreceiver/receiver.go#L139) is using a [JSON encoded response](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/master/receiver/signalfxreceiver/receiver.go#L198). 

I wonder if it is better if we move this middleware to a common package eg: `internal` and have it re-used wherever it's required(I think to wrap grpc-gateway exposed muxes for now) inside the collector as opposed to exposing it via `HTTPServerSettings`?

Reply:
Also, I just noticed that the contrib library is using HTTPServerSettings.ToServer. I suppose it will be difficult to have functions that work for all use-cases. Eg: signalfxreceiver is using a JSON encoded response.

I wonder if it is better if we move this middleware to a common package eg: internal and have it re-used wherever it's required(I think to wrap grpc-gateway exposed muxes for now) inside the collector as opposed to exposing it via HTTPServerSettings?